common/deadlock/deadlock.go

```diff
@@ -167,57 +173,32 @@ func (lc *loopContext) run(ctx context.Context) error {
 func (lc *loopContext) ping(ctx context.Context, pingables []pingable.Pingable) {
 	for _, pingable := range pingables {
 		for _, check := range pingable.GetPingChecks() {
-			select {
-			case lc.ch <- check:
-			case <-ctx.Done():
-				return
-			default:
-				// maybe add another worker if blocked
-				w := atomic.LoadInt32(&lc.workers)
-				if w < int32(lc.dd.config.MaxWorkersPerRoot()) && atomic.CompareAndSwapInt32(&lc.workers, w, w+1) {
-					lc.dd.loops.Go(lc.worker)
-				}
-				// blocking send
-				select {
-				case lc.ch <- check:
-				case <-ctx.Done():
-					return
-				}
-			}
+			lc.p.Do(func() { lc.check(ctx, check) })
 		}
 	}
 }
 
-func (lc *loopContext) worker(ctx context.Context) error {
-	for {
-		var check pingable.Check
-		select {
-		case check = <-lc.ch:
-		case <-ctx.Done():
-			return nil
-		}
-
-		lc.dd.logger.Debug("starting ping check", tag.Name(check.Name))
-		startTime := time.Now().UTC()
-
-		// Using AfterFunc is cheaper than creating another goroutine to be the waiter, since
-		// we expect to always cancel it. If the go runtime is so messed up that it can't
-		// create a goroutine, that's a bigger problem than we can handle.
-		t := time.AfterFunc(check.Timeout, func() {
-			if ctx.Err() != nil {
-				// deadlock detector was stopped
-				return
-			}
-			lc.dd.detected(check.Name)
-		})
-		newPingables := check.Ping()
-		t.Stop()
-		if len(check.MetricsName) > 0 {
-			lc.dd.metricsHandler.Timer(check.MetricsName).Record(time.Since(startTime))
+func (lc *loopContext) check(ctx context.Context, check pingable.Check) {
+	lc.dd.logger.Debug("starting ping check", tag.Name(check.Name))
+	startTime := time.Now().UTC()
+
+	// Using AfterFunc is cheaper than creating another goroutine to be the waiter, since
+	// we expect to always cancel it. If the go runtime is so messed up that it can't
+	// create a goroutine, that's a bigger problem than we can handle.
+	t := time.AfterFunc(check.Timeout, func() {
+		if ctx.Err() != nil {
+			// deadlock detector was stopped
+			return
 		}
+		lc.dd.detected(check.Name)
+	})
+	newPingables := check.Ping()
```

Comment:
I know you ported the logic as-is but is there any concern of blocking this goroutine on `Ping` when the context is canceled?

Reply:
Could be nice to be able to know that all goroutines are gathered when `run` returns.

Reply:
There's a comment in Stop().. I don't want to block because if there is a deadlock, that would propagate it further. I guess it could wait for some time and if it thinks one is still blocked, Stop() could return an error? Can you think of any way that it would make a practical difference?

Reply:
I don't think this is something worth blocking the PR for, you'd have to let `Ping` accept a context and that's a bigger change.

I generally would prefer in our codebase where something can block that we propagate whatever context we have.

Reply:
Sure, in general I agree, it's just that this is the deadlock detector.. it's the one component that we want to behave in a predictable way if there's a (potential) deadlock in something else. Anything else can just be oblivious to things blocking forever, and it would propagate deadlocks, but that's okay. This one, I would like Stop to reliably return even if some component has hit a (potential) deadlock. It can just be set up so that if Stop has been called, no more Ping checks are called, but the ones in progress can return.

Reply:
Looking again, there is a way to improve the stop propagation, not all the way through Ping, but it could stop the pools in dd.Stop instead of letting each worker goroutine do it. That would help if the pool is fully blocked and it can't even submit the root check. (This would be an extreme case but possibly it could happen if a deadlock spreads to all shards or something.) I'll do that.